File: python/controllers/file_controller.py
"""
File Controller Module

Responsible for defining the server's response to requests related to file operations.
Implements robust video streaming with byte range support and appropriate handling of
different file types (streaming vs. download).
"""
import os
import mimetypes
from flask import Blueprint, request, jsonify, send_file, Response, current_app, g
from werkzeug.utils import secure_filename
from models.file_model import FileModel

class FileController:
    """
    Controller for handling file-related operations including streaming and downloads.
    Uses a class-based approach for better organization and OOP principles.
    """
    
    def __init__(self, file_model=None):
        """
        Initialize with dependency injection for better testability.
        
        Args:
            file_model: The file model to use (defaults to FileModel if None)
        """
        self.file_model = file_model or FileModel()
        
    def upload_file(self):
        """
        Handles file upload.
        
        Returns:
            JSON response indicating success or error
        """
        if 'file' not in request.files:
            return jsonify({'error': 'No file part'}), 400
            
        file = request.files['file']
        
        if file.filename == '':
            return jsonify({'error': 'No selected file'}), 400
            
        # Get username from auth middleware via g
        username = g.user.get('username')
        
        try:
            # Secure the filename to prevent any malicious file paths
            filename = secure_filename(file.filename)
            
            # Determine file path
            upload_folder = current_app.config['UPLOAD_FOLDER']
            filepath = os.path.join(upload_folder, filename)
            
            # Save the file
            file.save(filepath)
            
            # Store file information in database
            file_info = self.file_model.create_file_record(
                filename=filename,
                filepath=filepath,
                mimetype=file.mimetype,
                size=os.path.getsize(filepath),
                uploaded_by=username
            )
            
            return jsonify({
                'message': 'File uploaded successfully',
                'fileId': file_info.id,
                'filename': file_info.filename
            }), 201
            
        except Exception as err:
            print(f"Error uploading file: {err}")
            return jsonify({'error': 'An error occurred during file upload'}), 500
    
    def get_file_list(self):
        """
        Gets a list of all files available to the user.
        
        Returns:
            JSON response with file list or error
        """
        try:
            # No longer require authenticated user for file listing 
            # username = g.user.get('username')  # removed to avoid missing user error
            
            # Get files (could filter by user permissions in a more complex implementation)
            files = self.file_model.get_all_files()
            
            return jsonify({
                'files': [file.to_dict() for file in files]
            }), 200
            
        except Exception as err:
            print(f"Error getting file list: {err}")
            return jsonify({'error': 'An error occurred while retrieving files'}), 500
    
    def get_file(self, file_id):
        """
        Gets a specific file by its ID.
        
        Args:
            file_id: ID of the file to retrieve
            
        Returns:
            File content with appropriate headers or error response
        """
        try:
            file_info = self.file_model.get_file_by_id(file_id)
            
            if not file_info:
                return jsonify({'error': 'File not found'}), 404
                
            filepath = file_info.filepath
            
            if not os.path.exists(filepath):
                return jsonify({'error': 'File not found on server'}), 404
                
            # Check if this is a streamable file type
            mimetype = file_info.mimetype or mimetypes.guess_type(filepath)[0]
            is_streamable = mimetype and (
                mimetype.startswith('video/') or 
                mimetype.startswith('audio/')
            )
            
            # Handle streaming for video/audio content
            if is_streamable:
                return self._stream_file(filepath, mimetype)
                
            # For other file types, return as attachment for download
            return send_file(
                filepath,
                mimetype=mimetype,
                as_attachment=True,
                attachment_filename=file_info.filename
            )
                
        except Exception as err:
            print(f"Error retrieving file: {err}")
            return jsonify({'error': 'An error occurred while retrieving the file'}), 500
    
    def _stream_file(self, filepath, mimetype):
        """
        Streams a file with byte range support.
        
        Args:
            filepath: Path to the file to stream
            mimetype: MIME type of the file
            
        Returns:
            Response object with appropriate headers for streaming
        """
        file_size = os.path.getsize(filepath)
        
        # Handle range requests for video/audio streaming
        range_header = request.headers.get('Range', None)
        
        if range_header:
            # Parse the range header
            byte_start, byte_end = 0, None
            if range_header.startswith('bytes='):
                range_str = range_header[6:].split('-')
                if range_str[0]:
                    byte_start = int(range_str[0])
                if len(range_str) > 1 and range_str[1]:
                    byte_end = int(range_str[1])
            
            if byte_end is None:
                byte_end = file_size - 1
                
            # Make sure we don't exceed the file size
            byte_end = min(byte_end, file_size - 1)
            
            # Calculate the content length
            content_length = byte_end - byte_start + 1
            
            # Create response headers
            headers = {
                'Content-Range': f'bytes {byte_start}-{byte_end}/{file_size}',
                'Accept-Ranges': 'bytes',
                'Content-Length': str(content_length),
                'Content-Type': mimetype
            }
            
            # Open the file at the specified position
            def generate():
                with open(filepath, 'rb') as f:
                    f.seek(byte_start)
                    data = f.read(content_length)
                    yield data
            
            return Response(generate(), 206, headers)
        else:
            # If no range header, return the entire file
            def generate():
                with open(filepath, 'rb') as f:
                    while True:
                        chunk = f.read(8192)  # Read in 8KB chunks
                        if not chunk:
                            break
                        yield chunk
            
            headers = {
                'Accept-Ranges': 'bytes',
                'Content-Length': str(file_size),
                'Content-Type': mimetype
            }
            
            return Response(generate(), 200, headers)
    
    def delete_file(self, file_id):
        """
        Deletes a file by its ID.
        
        Args:
            file_id: ID of the file to delete
            
        Returns:
            JSON response indicating success or error
        """
        try:
            # Get user info from auth middleware via g
            username = g.user.get('username')
            is_admin = g.user.get('is_admin', False)
            
            file_info = self.file_model.get_file_by_id(file_id)
            
            if not file_info:
                return jsonify({'error': 'File not found'}), 404
                
            # Check if the user has permission (owner or admin)
            if not is_admin and file_info.uploaded_by != username:
                return jsonify({'error': 'Permission denied'}), 403
                
            # Delete the file from storage
            if os.path.exists(file_info.filepath):
                os.remove(file_info.filepath)
                
            # Delete the file record from the database
            self.file_model.delete_file(file_id)
            
            return jsonify({'message': 'File deleted successfully'}), 200
            
        except Exception as err:
            print(f"Error deleting file: {err}")
            return jsonify({'error': 'An error occurred while deleting the file'}), 500

    def search_files(self):
        """
        Searches for files based on query parameters.
        
        Returns:
            JSON response with search results or error
        """
        try:
            query = request.args.get('query', '')
            
            files = self.file_model.search_files(query)
            
            return jsonify({
                'results': [file.to_dict() for file in files]
            }), 200
            
        except Exception as err:
            print(f"Error searching files: {err}")
            return jsonify({'error': 'An error occurred during file search'}), 500
        

# Create a Flask blueprint for file routes
file_bp = Blueprint('file', __name__)

# Create controller instance
file_controller = FileController()

# Define routes
@file_bp.route('/upload', methods=['POST'])
def upload():
    """Route for file upload"""
    return file_controller.upload_file()

@file_bp.route('/list', methods=['GET'])
def get_files():
    """Route to get all files"""
    return file_controller.get_file_list()

@file_bp.route('/<int:file_id>', methods=['GET'])
def get_file(file_id):
    """Route to get a specific file"""
    return file_controller.get_file(file_id)

@file_bp.route('/<int:file_id>', methods=['DELETE'])
def delete_file(file_id):
    """Route to delete a file"""
    return file_controller.delete_file(file_id)

@file_bp.route('/search', methods=['GET'])
def search():
    """Route to search files"""
    return file_controller.search_files()

File: js/controllers/fileController.js
/**
 * @module controller/fileController
 * Handles serving and storage of files to logged-in users
 */
const db = require('../database/db');
const path = require('path');
const File = require('../models/fileModel');
const fs = require('fs');

/**
 * Uploads a file and stores its metadata in the database.
 * @param {Object} req - The request object.
 * @param {Object} res - The response object.
 */
exports.uploadFile = async (req, res) => {   
    // Step one is capturing the file's metadata from the request
    const fileData = {
      file_name: req.file.file_name,            // req.file
      original_name: req.file.original_name,    // req.file
      uploader: req.user.username,              // req.user
      upload_date: new Date(),  // might need to handle this server-side instead?
      file_path: req.file.path,                 // req.file
      is_streamable: req.body.is_streamable ? 1 : 0
    };

    // now we have to handle the file's contents, which are a little trickier
    try {
      const file = await File.create(fileData);
      res.status(201).json({message: 'File uploaded successfully!', file });
    } catch (err) {
      console.error('Error uploading file:', err);
      res.status(500).json({error: 'Internal server error' });
    }
  };

/** Retrieves a list of files from the database.
  * @param {Object} req - The request object.
  * @param {Object} res - The response object.
  */
exports.listFiles = async (req, res) => {
  try {
    const files = await File.getAll();
    res.json(files);
  } catch (err) {
    console.error('Error listing files:', err);     // log error to console
    res.status(500).json({ error: 'Internal server error' });
  }
}

/**
 * Generates a list of matching files based on a supplied {keyword}
 * @param {Object} req - The request object, contains the keyword
 * @param {Object} res - The response object.
 */
exports.searchFiles = async (req, res) => {
  // users will enter a search keyword, we must pull it from the request
  const { keyword } = req.query;
  try {
    // try to search the Api Service by invoking fileModel's search method
    const files = await File.search(keyword);
    res.json(files);
  } catch (err) {
    console.error('Error performing search:', err);     // log error to console
    res.status(500).json({ error: 'Internal server error' });
  }
}


/**
 * Called when clicking on a hyperlink to a file.
 * 
 * If the file being requested is a *video*, then it will 
 * automatically be converted to a ByteStream so that it 
 * can be played clientside in-browser (or in future versions, apps)
 * 
 * Otherwise, we just serve the file to be downloaded.
 * @param {*} req 
 * @param {*} res 
 */
exports.getFileById = async (req, res) => {
  try {
    // error out if the file can't be found 
    const file = await File.findById(req.params.id);
    if (!file) {
=======
    const file = await File.findById(req.params.id);
    
    if (!file) {                        // Make sure the file exists first
>>>>>>> Stashed changes
      return res.status(404).json({error: 'File not found'});
    }
    
    // 
    const filePath = path.resolve(file.file_path);
    
    if (file.is_streamable) {           // Stream the file if it's compatible
      const stat = fs.statSync(filePath);
      const fileSize = stat.size;
      const range = req.headers.range;
      
      // if the file is valid, we'll need to funnel its bytes to the client
      if (range) { 
        // parse the file's bytes and turn them into a ByteStream
        const parts = range.replace(/bytes=/, '').split('-');
        const start = parseInt(parts[0], 10);
        const end = parts[1]
        ? parseInt(parts[1], 10)
          : fileSize - 1;
        const chunksize = end - start + 1;
        const fileStream = fs.createReadStream(filePath, { start, end });

        res.writeHead(206, {
          'Content-Range': `bytes ${start}-${end}/${fileSize}`,
          'Accept-Ranges': 'bytes',
          'Content-Length': chunksize,
          'Content-Type': 'video/mp4',
        });

        // now the stream is ready!
        fileStream.pipe(res);
      } else {
        res.writeHead(200, {
          'Content-Length': fileSize,
          'Content-Type': 'video/mp4'
        });
        // to send the file to the user, it's piped through an HTTP Response
        fs.createReadStream(filePath).pipe(res);
      }
    } 
    // If it's not a streamable video, then serve as download
    else {
      res.download(filePath, file.original_name, (err) => {
        if (err) {
          console.error('Error sending file:', err);
          res.status(500).json({ error: 'Internal server error' });
        }
      });
    }
  } catch (err) {
    console.error('Error retrieving file:', err);
    res.status(500).json({ error: 'Internal server error' });
  }
}